ouroboros-network-api/src/Ouroboros/Network/Block.hs

```diff
@@ -473,16 +473,13 @@ fromSerialised dec (Serialised payload) =
 --
 -- TODO: replace with encodeEmbeddedCBOR from cborg-0.2.4 once
 -- it is available, since that will be faster.
---
--- TODO: Avoid converting to a strict ByteString, as that requires copying O(n)
--- in case the lazy ByteString consists of more than one chunks.
 instance Serialise (Serialised a) where
   encode (Serialised bs) = mconcat [
         Enc.encodeTag 24
-      , Enc.encodeBytes (Lazy.toStrict bs)
+        encode bs
```

Comment:
Nice, so we avoid lazy to strict conversion, which indeed is costly.

The only problem with it is that we're modifying serialisation, which won't work across different versions (backwards compatibility).

I think we should:
* add `SerialisedV2` newtype wrapper,
* bump version of the `NodeToClient` mini-protocol
* use it in `LocalQueryLedger` mini-protocol if the negotiated version allows for it

An alternative is to modify the instance without changing the encoding (so it remains backwards compatible).

@nfrisby do you agree?  I think, this type is mostly used in `ouroboros-conesnsus`.

Reply:
FTR the ticket for making this instance lazy without any changes to the format (or rather, the deserialization part, which is the difficult one) is #5114 which we recently talked about in the context of incremental ledger serialization.

Also note that this instance (via `(un)wrapCBORinCBOR`) is used for sending eg txs/headers/blocks via N2C/N2N protocols, so all of these also would have to be patched for backwards-compatibility (across all implementations that do anything with txs/headers/blocks). Therefore,

> An alternative is to modify the instance without changing the encoding (so it remains backwards compatible).

sounds simpler to me.

Reply:
> The only problem with it is that we're modifying serialisation

Indeed, but that's also a desired feature IMO as it should prevent evaluating the entire bytestring in one chunk on the server; which can instead stream the response to clients. (I am really seeing this in the context of the state-query protocol).

I like the idea of making this version-specific; although the _version_ here should be driven by the NodeToClient version in the case of the state query protocols. I believe the Serialised type is also used elsewhere, where version-constaints may be different but all that can very likely be resolved through a type-class or a type family.

Reply:
From the RFC (section `3.4.5.1`):

> Sometimes it is beneficial to carry an embedded CBOR data item that is not meant to be decoded immediately at the time the enclosing data item is being decoded. Tag number 24 (CBOR data item) can be used to tag the embedded byte string as a single data item encoded in CBOR format. Contained items that aren't byte strings are invalid.

It seems what you're proposing in this patch is not a valid CBOR, see [`encodeChunked`](https://hackage.haskell.org/package/serialise-0.2.6.1/docs/src/Codec.Serialise.Class.html#encodeChunked).

Reply:
I don't think you're reading this right. `encodeChunked` produces a single data item: a bytestring (major type 2).

Whether it is indefinite or finite doesn't really change the fact that it's a single CBOR data item.

Reply:
Yes, you're right.

I think we all agree on adding `SerialisedV2` to `ouroboros-newtork` and a new `NodeToClientVersion`.